src/irr/video/COpenGLRenderpassIndependentPipeline.h

```diff
@@ -53,9 +67,57 @@ class COpenGLRenderpassIndependentPipeline : public IGPURenderpassIndependentPip
             m_stagePresenceMask |= (static_cast<uint32_t>(present) << i);
         }
     }
+    ~COpenGLRenderpassIndependentPipeline()
+    {
+        //shader programs can be shared so all names can be freed by any thread
+        for (GLuint n : *m_GLnames)
+            if (n!=0u)
+                COpenGLExtensionHandler::extGlDeleteProgram(n);
+        _IRR_ALIGNED_FREE(m_uniformValues);
+    }
 
     uint32_t getStagePresenceMask() const { return m_stagePresenceMask; }
 
+    GLuint getShaderGLnameForCtx(uint32_t _stageIx, uint32_t _ctxID) const
+    {
+        const uint32_t name_ix = _ctxID*SHADER_STAGE_COUNT + _stageIx;
+
+        if ((*m_GLnames)[name_ix])
+            return (*m_GLnames)[name_ix];
+        else if ((*m_shaderBinaries)[_stageIx].binary)
+        {
+            const auto& bin = (*m_shaderBinaries)[_stageIx];
+
+            const GLuint GLname = COpenGLExtensionHandler::extGlCreateProgram();
+            COpenGLExtensionHandler::extGlProgramBinary(GLname, bin.format, bin.binary->data(), bin.binary->size());
+            (*m_GLnames)[_ctxID] = GLname;
+            return GLname;
+        }
+
+        std::tie((*m_GLnames)[name_ix], (*m_shaderBinaries)[_stageIx]) = 
+            static_cast<const COpenGLSpecializedShader*>(m_shaders[_stageIx].get())->compile(static_cast<const COpenGLPipelineLayout*>(getLayout()));
+        return (*m_GLnames)[name_ix];
+    }
+
+    //assumes GL shader object for _stageIx for _ctxID was already created (ctor or getShaderGLnameForCtx() )!
+    const GLint* getUniformLocationsForStage(uint32_t _stageIx, uint32_t _ctxID) const
+    {
+        if (m_uniformLocations[_stageIx])
+            return m_uniformLocations[_stageIx]->data();
+
+        const COpenGLSpecializedShader* glshdr = static_cast<const COpenGLSpecializedShader*>(m_shaders[_stageIx].get());
+        auto uniforms_rng = glshdr->getUniforms();
+        GLuint GLname = (*m_GLnames)[_ctxID*SHADER_STAGE_COUNT + _stageIx];
+
+        m_uniformLocations[_stageIx] = core::make_refctd_dynamic_array<decltype(m_uniformLocations[_stageIx])>(uniforms_rng.length());
+        for (size_t i = 0ull; i < uniforms_rng.length(); ++i)
+            (*m_uniformLocations[_stageIx])[i] = COpenGLExtensionHandler::extGlGetUniformLocation(GLname, uniforms_rng.begin()[i].m.name.c_str());
+
+        return m_uniformLocations[_stageIx]->data();
+    }
```

Comment:
because you're using separable programs, the uniform locations can be precomputed (will be identical for all usages across all threads and pipelines) for a specialized shader (from a single compilation) and "cached" inside of the specialized shader

Reply:
they're precomputed once for all threads, however u're right that those can be stored in COpenGLSpecializedShader, i'll change it